### PR TITLE
reduce allocations for string(::IEEEFloat)

### PR DESCRIPTION
**Pull request by oscardssmith**
```
julia> @time string(1.0)
  0.000004 seconds (3 allocations: 432 bytes) # before
  0.000007 seconds (3 allocations: 112 bytes) # after
julia> io = IOBuffer();
julia> @time show(io, 1.0)
  0.000014 seconds (3 allocations: 432 bytes) # before
  0.000012 seconds (1 allocation: 48 bytes) # after

```
mitigates https://github.com/JuliaLang/julia/issues/57976

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Comment by bvdmitri** [comment]
Can this PR also fix `Printf.format` in a similar manner? Currently it does allocate an intermediate array too and works almost in exactly same way as the `show` method. With the `Printf.format` though by using some internal API it is possible to do non-allocating printing of a number
```julia
julia> buf = Vector{UInt8}(undef, 10)
10-element Vector{UInt8}:
 0x69
 0x09
 0x00
 0x00
 0x00
 0x00
 0x00
 0x00
 0x6a
 0x09

julia> format = Printf.format"%.4f"
Printf.Format{Base.CodeUnits{UInt8, String}, Tuple{Printf.Spec{Val{'f'}}}}(UInt8[0x25, 0x2e, 0x34, 0x66], UnitRange{Int64}[1:0, 5:4], (%.4f,), 1)

julia> @time Printf.format(buf, 1, format, 1.0)
  0.000016 seconds
7
```

However, the signature of `format` is constrained to `Vector` and cannot use the `Memory`

**Comment by oscardssmith** [comment]
I think `printf` should be left to a separate PR. The last thing to do here, is I would appreciate someone double checking that the numbers I've put down for `neededfloatdigits` are correct (and ideally find the formula that generates them).  I believe these are the right numbers, but my search for them wasn't exactly principled)

File: base/ryu/Ryu.jl [hunk 1]
```diff
@@ -111,16 +120,16 @@ end
 
 function Base.show(io::IO, x::T, forceuntyped::Bool=false, fromprint::Bool=false) where {T <: Base.IEEEFloat}
     compact = get(io, :compact, false)::Bool
-    buf = Base.StringVector(neededdigits(T))
+    buf = Memory{UInt8}(undef, neededfloatdigits(T))
```
Comment (vtjnash) [inline base/ryu/Ryu.jl:123]: Extra bytes are needed (for "Float16()") if "typed && T isa Float16"

File: base/ryu/Ryu.jl [hunk 2]
```diff
@@ -17,6 +17,15 @@ neededdigits(::Type{Float64}) = 309 + 17
 neededdigits(::Type{Float32}) = 39 + 9 + 2
 neededdigits(::Type{Float16}) = 9 + 5 + 9
 
+"""
+    Ryu.neededfloatdigits(T)
+
+Number of digits necessary to represent type `T` in shortest precision.
+"""
+neededfloatdigits(::Type{Float64}) = 23
```
Comment (vtjnash) [inline base/ryu/Ryu.jl:25]: "If an IEEE 754 double-precision number is converted to a decimal string with at least 17 significant digits, and then converted back to double-precision representation, the final result must match the original number"

17 + decimal point + sign bit + "e" + second sign bit + max exponent (3) = 24?

https://en.wikipedia.org/wiki/Double-precision_floating-point_format

File: base/ryu/Ryu.jl [hunk 3]
```diff
@@ -17,6 +17,15 @@ neededdigits(::Type{Float64}) = 309 + 17
...
+neededfloatdigits(::Type{Float32}) = 15
```
Comment (vtjnash) [inline base/ryu/Ryu.jl:26]: 9 + decimal point + sign bit + "f" + second sign bit + max exponent (2) = 15

File: base/ryu/Ryu.jl (hunk 1 above)
Comment (oscardssmith) [reply base/ryu/Ryu.jl:123]: I included this in `neededfloatdigits(::Float16)`, but I guess it would make sense to separate it so we can allocate less when `!typed`

File: base/ryu/Ryu.jl [hunk 4]
```diff
@@ -17,6 +17,15 @@ neededdigits(::Type{Float64}) = 309 + 17
...
+neededfloatdigits(::Type{Float16}) = 18
```
Comment (vtjnash) [inline base/ryu/Ryu.jl:27]: 5 + decimal point + sign bit + "e" + second sign bit + max exponent (2) = 11
```
julia> extrema(i->length(string(reinterpret(Float16, i))), typemin(UInt16):typemax(UInt16))
(3, 11)
```

(so maximum = 20 with typed=true)

File: base/ryu/Ryu.jl (hunk 3 above)
Comment (vtjnash) [reply base/ryu/Ryu.jl:26]:
```
julia> extrema(i->length(string(reinterpret(Float32, i))), typemin(UInt32):typemax(UInt32))
(3, 15)
```

**Comment by KristofferC** [comment]
Would be good with tests for the cases where the constants were wrong since that seems like good edge casea to test.

**Comment by oscardssmith** [comment]
tests added.

**Comment by KristofferC** [comment]
```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-x64-5.0/build/default-macmini-x64-5-0/julialang/julia-master/julia-a03f845f70/share/julia/test/ryu.jl:310
  | Expression: "-1.00000075f-36" == Ryu.writeshortest(-1.00000075f-36)
  | Evaluated: "-1.00000075f-36" == "-1.00000075e-36"
  |  
```

So, I guess, no?

**Comment by oscardssmith** [comment]
oops. Tests now look a lot happier.

**Comment by oscardssmith** [comment]
ready to merge now?

File: base/ryu/Ryu.jl [hunk 5]
```diff
@@ -17,6 +17,15 @@ neededdigits(::Type{Float64}) = 309 + 17
...
 neededdigits(::Type{Float16}) = 9 + 5 + 9
 
+"""
+    Ryu.neededfloatdigits(T)
+
+Number of digits necessary to represent type `T` in shortest precision.
+"""
+neededfloatdigits(::Type{Float64}) = 24
+neededfloatdigits(::Type{Float32}) = 15
+neededfloatdigits(::Type{Float16}) = 20
```
Comment (JeffBezanson) [inline base/ryu/Ryu.jl:27]: Should this be 11?

File: test/ryu.jl [hunk 6]
```diff
@@ -368,6 +372,9 @@ end
 #     return fails / (fails + success)
 # end
 
+@testset "Regression" begin
```
Comment (KristofferC) [inline test/ryu.jl:375]: I don't know if a test set called "regression" with a single test in it is that useful.

File: base/ryu/Ryu.jl (hunk 5 above)
Comment (oscardssmith) [reply base/ryu/Ryu.jl:27]: no. since we print the type name

**Comment by oscardssmith** [comment]
merge?

File: base/ryu/Ryu.jl [hunk 7]
```diff
@@ -111,16 +120,16 @@ end
...
     pos = writeshortest(buf, 1, x, false, false, true, -1,
         (x isa Float32 && !fromprint) ? UInt8('f') : UInt8('e'), false, UInt8('.'), typed, compact)
-    write(io, resize!(buf, pos - 1))
+    write(io, view(buf, 1:pos - 1))
     return
 end
 
 function Base.string(x::T) where {T <: Base.IEEEFloat}
-    buf = Base.StringVector(neededdigits(T))
+    buf = Base.StringVector(neededfloatdigits(T))
```
Comment (LilithHafner) [inline base/ryu/Ryu.jl:132]: This could be 11 instead of 20 for Float16, but no big deal.

File: test/ryu.jl [hunk 8]
```diff
@@ -418,6 +422,9 @@ end
...
+    @test "-0.00010014" == Ryu.writeshortest(Float16(-0.00010014)) #longest Float16
+end
```
Comment (oscardssmith) [inline test/ryu.jl]:
```suggestion
@test "-0.00010014" == Ryu.writeshortest(Float16(-0.00010014)) #longest Float16
```

File: base/ryu/Ryu.jl (hunk 3 above)
Comment (LilithHafner) [reply base/ryu/Ryu.jl]:
```julia-repl
julia> length(sprint(show, -0.000110974266f0))
17
```

**Review by LilithHafner — DISMISSED** [review]
Sometimes writeshortest displays in a fixed point format that would have been shorter than or as short as scientific noration, assuming that the fixed point format will not have `f0` appended to it. However, this assumption is not true when we need to display typeinfo. See https://github.com/JuliaLang/julia/pull/57977#discussion_r2233918567 and CI for examples of this. Adding 2 to the Float32 case should be sufficient to address this, I suspect, but we should be careful.

File: base/ryu/Ryu.jl (hunk 3 above)
Comment (LilithHafner) [reply base/ryu/Ryu.jl]:
```suggestion
neededfloatdigits(::Type{Float32}) = 17
```
Add 2 for `0f` (see https://github.com/JuliaLang/julia/pull/57977#pullrequestreview-3059472408)

File: base/ryu/Ryu.jl [hunk 9]
```diff
@@ -23,7 +23,7 @@ neededdigits(::Type{Float16}) = 9 + 5 + 9
 Number of digits necessary to represent type `T` in shortest precision.
 """
 neededfloatdigits(::Type{Float64}) = 24
-neededfloatdigits(::Type{Float32}) = 15
+neededfloatdigits(::Type{Float32}) = 17
```
Comment (KristofferC) [inline base/ryu/Ryu.jl:26]: I think always when CI passed but we still need to bump this number, a test should be added for the case that required more digits since it clearly is missing some coverage?

Comment (LilithHafner) [reply base/ryu/Ryu.jl:26]: CI didn't pass but I agree that this should get a specific test.

**Comment by bvdmitri** [comment]
bump

Just to update on my previously opened issue https://github.com/JuliaLang/julia/issues/57976, which this PR attempts to mitigate. I think the discussion around `neededfloatdigits` isn't the part of the original issue, because I truly believe that printing a `Float64` to a pre-allocated buffer should not allocate anything.
